query/src/main/scala/filodb/query/exec/ExecPlan.scala

```diff
@@ -111,82 +105,92 @@ trait ExecPlan extends QueryCommand {
-            srv
-          case rv: RangeVector =>
-            // materialize, and limit rows per RV
-            val srv = SerializableRangeVector(rv, builder, recSchema, printTree(false))
-            numResultSamples += srv.numRowsInt
-            // fail the query instead of limiting range vectors and returning incomplete/inaccurate results
-            if (enforceLimit && numResultSamples > limit)
-              throw new BadQueryException(s"This query results in more than $limit samples. " +
-                s"Try applying more filters or reduce time range.")
-            srv
+      doExecute(source, queryConfig)
+    }
+
+    // Step 2: Set up transformers and loop over all rangevectors, creating the result
+    def step2(res: ExecResult) = res.schema.map { resSchema =>
+      // It is possible a null schema is returned (due to no time series). In that case just return empty results
+      val resultTask = if (resSchema == ResultSchema.empty) {
+        qLogger.debug(s"Empty plan $this, returning empty results")
+        Task.eval(QueryResult(id, resSchema, Nil))
+      } else {
+        val finalRes = allTransformers.foldLeft((res.rvs, resSchema)) { (acc, transf) =>
+          qLogger.debug(s"queryId: ${id} Setting up Transformer ${transf.getClass.getSimpleName} with ${transf.args}")
+          (transf.apply(acc._1, queryConfig, limit, acc._2), transf.schema(acc._2))
         }
-        .take(limit)
-        .toListL
-        .map { r =>
-          val numBytes = builder.allContainers.map(_.numBytes).sum
-          SerializableRangeVector.queryResultBytes.record(numBytes)
-          if (numBytes > 5000000) {
-            // 5MB limit. Configure if necessary later.
-            // 250 RVs * (250 bytes for RV-Key + 200 samples * 32 bytes per sample)
-            // is < 2MB
-            qLogger.warn(s"queryId: ${id} result was " +
-              s"large size ${numBytes}. May need to tweak limits. " +
-              s"ExecPlan was: ${printTree()} " +
-              s"Limit was: ${limit}")
+        val recSchema = SerializableRangeVector.toSchema(finalRes._2.columns, finalRes._2.brSchemas)
+        val builder = SerializableRangeVector.newBuilder()
+        var numResultSamples = 0 // BEWARE - do not modify concurrently!!
+        finalRes._1
+          .map {
+            case srv: SerializableRangeVector =>
+              numResultSamples += srv.numRowsInt
+              // fail the query instead of limiting range vectors and returning incomplete/inaccurate results
+              if (enforceLimit && numResultSamples > limit)
+                throw new BadQueryException(s"This query results in more than $limit samples. " +
+                  s"Try applying more filters or reduce time range.")
+              srv
+            case rv: RangeVector =>
+              // materialize, and limit rows per RV
+              val srv = SerializableRangeVector(rv, builder, recSchema, printTree(false))
```

Comment:
I don't get the constant urge to verify what thread different things are running on.  There has been no evidence that this part of the code causes any problems.  Let's add this only if we have problems in this area.

Reply:
I have mentioned this before - our use of monix framework doesnt help with thread pool standing out for methods. The assertion indeed helps ensure that code is running on the intended threads. This specific one is important since the locks are acquired/released here.